bigquery/src/main/java/org/apache/iceberg/gcp/bigquery/BigQueryMetastoreCatalog.java

```diff
@@ -0,0 +1,400 @@
+        .map(
+            table -> TableIdentifier.of(namespace.level(0), table.getTableReference().getTableId()))
+        .collect(ImmutableList.toImmutableList());
+  }
+
+  @Override
+  public boolean dropTable(TableIdentifier identifier, boolean purge) {
+    try {
+      TableOperations ops = newTableOps(identifier);
+      TableMetadata lastMetadata = ops.current();
+
+      client.deleteTable(toTableReference(identifier));
+
+      if (purge && lastMetadata != null) {
+        CatalogUtil.dropTableData(ops.io(), lastMetadata);
+      }
+    } catch (NoSuchTableException e) { // Not catching a NoSuchIcebergTableException on purpose
+      return false; // The documentation says just return false in this case
+    }
+    return true;
+  }
+
+  @Override
+  public void renameTable(TableIdentifier from, TableIdentifier to) {
+    validateNamespace(from.namespace());
+    validateNamespace(to.namespace());
+
+    try {
+      client.getDataset(toDatasetReference(to.namespace()));
+      client.getDataset(toDatasetReference(from.namespace()));
+    } catch (NoSuchNamespaceException e) {
+      throw e;
+    }
+
+    TableReference fromTableRef = toTableReference(from);
+    TableReference toTableRef = toTableReference(to);
+
+    // Check if the source table exists
+    try {
+      client.getTable(fromTableRef);
+    } catch (NoSuchTableException e) {
+      throw new NoSuchTableException("Table does not exist: %s", from.name());
+    }
+
+    // Check if the destination table already exists
+    try {
+      client.getTable(toTableRef);
+      // If getTable succeeds, the destination table exists
+      throw new AlreadyExistsException("Table already exists: %s", to.name());
+    } catch (NoSuchTableException e) {
+      // Destination table does not exist, proceed with throwing exception
+    }
+
+    if (!from.namespace().equals(to.namespace())) {
+      throw new ValidationException("New table name must be in the same namespace");
+    }
+
+    // TODO(b/354981675): Enable once supported by the API.
+    throw new UnsupportedOperationException("Table rename operation is unsupported.");
+  }
+
+  @Override
+  public void createNamespace(Namespace namespace, Map<String, String> metadata) {
+    Dataset builder = new Dataset();
+    DatasetReference datasetReference = toDatasetReference(namespace);
+    builder.setLocation(this.location);
+    builder.setDatasetReference(datasetReference);
+    builder.setExternalCatalogDatasetOptions(
+        BigQueryMetastoreUtils.createExternalCatalogDatasetOptions(
+            datasetReference.getDatasetId(), metadata));
+
+    client.createDataset(builder);
+  }
+
+  /**
+   * Since this catalog only supports one-level namespaces, it always returns an empty list unless
+   * passed an empty namespace to list all namespaces within the catalog.
+   */
+  @Override
+  public List<Namespace> listNamespaces(Namespace namespace) {
```

Comment:
this needs to check whether the namespace exists or not and throw a `NoSuchNamespaceException` if the namespace doesn't exist. I also realized we don't have a test for this in `CatalogTests`, so I opened https://github.com/apache/iceberg/pull/12873

Reply:
BigQuery MetaStore is the first catalog that supports this. I am proud of it 😃